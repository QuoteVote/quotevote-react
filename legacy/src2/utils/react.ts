import React, { 
  useState, 
  useEffect, 
  useRef, 
  ReactNode, 
  ChangeEvent, 
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
} from 'react';

export type { 
  ReactNode,
  ChangeEvent,
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
};

export { 
  React as default,
  useState,
  useEffect,
  useRef
};

// Add type declarations for common event handlers
export type InputChangeEvent = ChangeEvent<HTMLInputElement>;
export type TextAreaChangeEvent = ChangeEvent<HTMLTextAreaElement>;
export type ButtonClickEvent = MouseEvent<HTMLButtonElement>;
export type SetState<T> = Dispatch<SetStateAction<T>>;

// Add type declarations for common component props
export interface BaseProps {
  className?: string;
  style?: React.CSSProperties;
  children?: ReactNode;
}

export interface InputProps extends BaseProps {
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
  disabled?: boolean;
} 