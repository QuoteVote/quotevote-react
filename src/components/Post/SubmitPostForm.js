import React from 'react'
import TextField from '@material-ui/core/TextField'
import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete'
import { makeStyles } from '@material-ui/core/styles'
import {
  CircularProgress,
  Divider, FormControl, Grid, InputBase, Typography,
} from '@material-ui/core'
import { useForm } from 'react-hook-form'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import PropTypes from 'prop-types'
import { useMutation } from '@apollo/react-hooks'
import { useDispatch } from 'react-redux'
import CardBody from '../../mui-pro/Card/CardBody'
import Card from '../../mui-pro/Card/Card'
import Button from '../../mui-pro/CustomButtons/Button'
import SubmitPostAlert from './SubmitPostAlert'
import { SET_SELECTED_POST } from '../../store/ui'
import { CREATE_GROUP, SUBMIT_POST } from '../../graphql/mutations'

const useStyles = makeStyles({
  title: {
    fontSize: '25px',
    font: 'League Spartan',
    fontWeight: 'bold',
  },
  text: {
    marginTop: 20,
    fontSize: '20px',
  },
  group: {
    fontSize: '20px',
  },
  button: {
    backgroundColor: '#00CF6E',
  },
})

const filter = createFilterOptions()

function SubmitPostForm({ options = [], user }) {
  const classes = useStyles()
  const dispatch = useDispatch()
  const {
    register, handleSubmit, errors, reset,
  } = useForm()
  const [submitPost, { loading }] = useMutation(SUBMIT_POST)
  const [createGroup, { loading: loadingGroup }] = useMutation(CREATE_GROUP)
  const [selectedGroup, setSelectedGroup] = React.useState(null)
  const [privacy, setPrivacy] = React.useState('public')
  const [isNewGroup, setIsNewGroup] = React.useState(false)
  const [error, setError] = React.useState(null)
  const handleVisibilityChange = (event) => {
    setPrivacy(event.target.value)
  }
  const onSubmit = async (values) => {
    const { title, text, group } = values
    try {
      let newGroup
      if (isNewGroup) {
        newGroup = await createGroup({
          variables: {
            group: {
              creatorId: user._id,
              title: group,
              description: `Description for: ${group} group`,
              privacy,
            },
          },
        })
      }
      const postGroupId = isNewGroup ? newGroup.data.createGroup._id : selectedGroup._id
      const submitResult = await submitPost({
        variables: {
          post: {
            userId: user._id,
            text,
            title,
            groupId: postGroupId,
          },
        },
      })
      const { _id, url } = submitResult.data.addPost
      dispatch(SET_SELECTED_POST(_id))
      setShareableLink(url)
      setShowAlert(true)
    } catch (err) {
      setError(err)
      setShowAlert(true)
    }
  }

  const [shareableLink, setShareableLink] = React.useState('')
  const [showAlert, setShowAlert] = React.useState(false)
  const hideAlert = () => {
    setShowAlert(false)
    setShareableLink('')
    setIsNewGroup(false)
    setSelectedGroup(null)
    reset()
  }
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {showAlert && (
        <SubmitPostAlert
          hideAlert={hideAlert}
          shareableLink={shareableLink}
          error={error}
        />
      )}
      <Card>
        <CardBody>
          <InputBase
            fullWidth
            id="title"
            className={classes.title}
            placeholder="Enter Title Here"
            name="title"
            inputRef={register({
              required: 'Title is required',
            })}
            required
            error={errors.title}
            helperText={errors.title && errors.title.message}
          />
          <Divider />
          <InputBase
            id="text"
            placeholder="Input text to submit post"
            className={classes.text}
            multiline
            fullWidth
            name="text"
            inputRef={register({
              required: 'Post is required',
            })}
            required
            error={errors.text}
          />
          <Divider />

          <Autocomplete
            value={selectedGroup}
            onChange={(event, newValue) => {
              if (typeof newValue === 'string') {
                setSelectedGroup({
                  title: newValue,
                })
              } else if (newValue && newValue.inputValue) {
                // Create a new value from the user input
                setSelectedGroup({
                  title: newValue.inputValue,
                })
              } else {
                setSelectedGroup(newValue)
              }

              if (newValue) {
                const checkGroup = (groupOption) => groupOption.title === newValue.title
                const hidePrivacyOption = options.some(checkGroup)
                setIsNewGroup(!hidePrivacyOption)
              }
            }}
            filterOptions={(groupOptions, params) => {
              const filtered = filter(groupOptions, params)

              // Suggest the creation of a new value
              if (params.inputValue !== '') {
                filtered.push({
                  inputValue: params.inputValue,
                  title: `Add "${params.inputValue}"`,
                })
                setIsNewGroup(true)
              } else {
                setIsNewGroup(false)
              }

              return filtered
            }}
            selectOnFocus
            clearOnBlur
            handleHomeEndKeys
            id="group"
            options={options}
            getOptionLabel={(option) => {
              // Value selected with enter, right from the input
              if (typeof option === 'string') {
                return option
              }
              // Add "xxx" option created dynamically
              if (option.inputValue) {
                return option.inputValue
              }
              // Regular option
              return option.title
            }}
            renderOption={(option) => option.title}
            renderInput={(params) => (
              <TextField
                {...params}
                label="Select a group"
                name="group"
                id="group"
                inputRef={register({
                  required: 'Group is required',
                })}
              />
            )}
            fullWidth
          />

          <Typography>
            You can create private or public groups of people to shared your information.
          </Typography>

          {isNewGroup && (
            <FormControl component="fieldset">
              <RadioGroup
                aria-label="Group"
                name="groupVisibility"
                value={privacy}
                onChange={handleVisibilityChange}
              >
                <FormControlLabel value="public" control={<Radio />} label="Public" />
                <FormControlLabel value="private" control={<Radio />} label="Private" />
              </RadioGroup>
            </FormControl>
          )}

          <Grid
            container
            direction="row"
            justify="flex-end"
            alignItems="flex-end"
          >
            <Button
              id="submit-button"
              type="submit"
              variant="contained"
              size="large"
              className={classes.button}
              disabled={loadingGroup || loading}
            >
              Submit
              {(loading || loadingGroup) && <CircularProgress size={20} style={{ marginLeft: 5 }} />}

            </Button>
          </Grid>

        </CardBody>
      </Card>
    </form>
  )
}
SubmitPostForm.propTypes = {
  options: PropTypes.object,
  user: PropTypes.object,
}

export default SubmitPostForm
