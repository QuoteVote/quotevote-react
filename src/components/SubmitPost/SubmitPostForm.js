import React from 'react'
import TextField from '@material-ui/core/TextField'
import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete'
import { makeStyles } from '@material-ui/core/styles'
import {
  CircularProgress,
  Divider, FormControl, Grid, InputBase, Typography, IconButton,
} from '@material-ui/core'
import { useForm } from 'react-hook-form'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import PropTypes from 'prop-types'
import Mercury from '@postlight/mercury-parser'
import { useMutation } from '@apollo/react-hooks'
import { useDispatch } from 'react-redux'
import { isEmpty } from 'lodash'
import CardBody from '../../mui-pro/Card/CardBody'
import Button from '../../mui-pro/CustomButtons/Button'
import SubmitPostAlert from './SubmitPostAlert'
import { SET_SELECTED_POST } from '../../store/ui'
import { CREATE_GROUP, SUBMIT_POST } from '../../graphql/mutations'

const useStyles = makeStyles({
  root: {
    width: 520,
  },
  title: {
    color: '#00CF6E',
    fontSize: 30,
    margin: 'auto',
    paddingLeft: 20,
  },
  exit: {
    color: '#00CF6E',
    fontSize: 30,
    float: 'right',
  },
  text: {
    marginTop: 20,
    fontSize: '20px',
  },
  input: {
    margin: '20px 0px 0px 0px',
  },
  group: {
    margin: '20px 0px',
  },
  groupInput: {
    backgroundColor: 'rgb(160, 243, 204, 0.6)',
    width: 180,
    marginLeft: 20,
  },
  label: {
    color: '#00CF6E',
  },
  button: {
    backgroundColor: '#75E2AF',
    fontSize: 20,
  },
})

const filter = createFilterOptions()

function SubmitPostForm({ options = [], user, setOpen }) {
  const classes = useStyles()
  const dispatch = useDispatch()
  const {
    register, handleSubmit, errors, reset,
  } = useForm()
  const [submitPost, { loading }] = useMutation(SUBMIT_POST)
  const [createGroup, { loading: loadingGroup }] = useMutation(CREATE_GROUP)
  const [selectedGroup, setSelectedGroup] = React.useState(null)
  const [privacy, setPrivacy] = React.useState('public')
  const [isNewGroup, setIsNewGroup] = React.useState(false)
  const [error, setError] = React.useState(null)
  const handleVisibilityChange = (event) => {
    setPrivacy(event.target.value)
  }

  const onSubmit = async (values) => {
    const { title, text, group } = values
    try {
      let newGroup
      if (isNewGroup) {
        newGroup = await createGroup({
          variables: {
            group: {
              creatorId: user._id,
              title: group,
              description: `Description for: ${group} group`,
              privacy,
            },
          },
        })
      }
      const postGroupId = isNewGroup ? newGroup.data.createGroup._id : selectedGroup._id
      const submitResult = await submitPost({
        variables: {
          post: {
            userId: user._id,
            text,
            title,
            groupId: postGroupId,
          },
        },
      })
      const { _id, url } = submitResult.data.addPost
      dispatch(SET_SELECTED_POST(_id))
      setShareableLink(url)
      setShowAlert(true)
    } catch (err) {
      setError(err)
      setShowAlert(true)
    }
  }

  const [shareableLink, setShareableLink] = React.useState('')
  const [showAlert, setShowAlert] = React.useState(false)
  const hideAlert = () => {
    setShowAlert(false)
    setShareableLink('')
    setIsNewGroup(false)
    setSelectedGroup(null)
    reset()
  }
  const [value, setValue] = React.useState({ title: '', content: '' })
  const [isPasting, setPasting] = React.useState(false)

  const handleTitleChange = (event) => {
    event.persist()
    setValue({ ...value, title: event.target.value })
  }

  const handleContentChange = (event) => {
    event.persist()
    const contentValue = event.target.value
    const validURL = /^(?:http(s)?:\/\/)([\w.-])+(?:[\w.-]+)+([\w\-._~:/?#[\]@!$&'()*+,;=.])+$/
    setValue({ ...value, content: contentValue })
    if (isPasting && contentValue.match(validURL)) {
      Mercury.parse(`https://cors-anywhere.herokuapp.com/${contentValue}`, { contentType: 'markdown' })
        .then((result) => {
          if (isEmpty(result.content)) {
            setError('Could not extract site content.')
            setShowAlert(true)
          } else {
            setValue({ title: result.title, content: result.content })
          }
        })
    }
    setPasting(false)
  }

  const handlePaste = () => {
    setPasting(true)
  }

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {showAlert && (
        <SubmitPostAlert
          hideAlert={hideAlert}
          shareableLink={shareableLink}
          error={error}
          setShowAlert={setShowAlert}
          setOpen={setOpen}
        />
      )}
      <CardBody className={classes.root}>
        <Grid
          container
          direction="row"
        >
          <Typography className={classes.title} variant="body2">Create a Post</Typography>
          <IconButton className={classes.exit} onClick={() => setOpen(false)}>X</IconButton>
        </Grid>
        <InputBase
          className={classes.input}
          fullWidth
          id="title"
          placeholder="Enter Title"
          value={value.title}
          onChange={(event) => { handleTitleChange(event) }}
          name="title"
          inputRef={register({
            required: 'Title is required',
          })}
          required
          error={errors.title}
        />
        <Divider />
        <InputBase
          className={classes.input}
          id="text"
          placeholder="Enter text or URL here"
          value={value.content}
          onChange={(event) => { handleContentChange(event) }}
          onPaste={handlePaste}
          multiline
          fullWidth
          name="text"
          inputRef={register({
            required: 'Post is required',
          })}
          required
          error={errors.text}
        />
        <Divider />
        <Grid
          container
          direction="row"
          justify="flex-start"
          alignItems="center"
          className={classes.group}
        >
          <Typography>
            Who can see your post
          </Typography>

          <Autocomplete
            variant="outlined"
            size="small"
            className={classes.groupInput}
            value={selectedGroup}
            onChange={(event, newValue) => {
              if (typeof newValue === 'string') {
                setSelectedGroup({
                  title: newValue,
                })
              } else if (newValue && newValue.inputValue) {
                // Create a new value from the user input
                setSelectedGroup({
                  title: newValue.inputValue,
                })
              } else {
                setSelectedGroup(newValue)
              }

              if (newValue) {
                const checkGroup = (groupOption) => groupOption.title === newValue.title
                const hidePrivacyOption = options.some(checkGroup)
                setIsNewGroup(!hidePrivacyOption)
              }
            }}
            filterOptions={(groupOptions, params) => {
              const filtered = filter(groupOptions, params)

              // Suggest the creation of a new value
              if (params.inputValue !== '') {
                filtered.push({
                  inputValue: params.inputValue,
                  title: `Add "${params.inputValue}"`,
                })
                setIsNewGroup(true)
              } else {
                setIsNewGroup(false)
              }

              return filtered
            }}
            selectOnFocus
            clearOnBlur
            handleHomeEndKeys
            id="hashtag"
            options={options}
            getOptionLabel={(option) => {
              // Value selected with enter, right from the input
              if (typeof option === 'string') {
                return option
              }
              // Add "xxx" option created dynamically
              if (option.inputValue) {
                return option.inputValue
              }
              // Regular option
              return option.title
            }}
            renderOption={(option) => option.title}
            renderInput={(params) => (
              <TextField
                variant="outlined"
                className={classes.label}
                {...params}
                label="Select a hashtag"
                name="hashtag"
                id="hashtag"
                inputRef={register({
                  required: 'hashtag is required',
                })}
              />
            )}
          />

          {isNewGroup && (
            <FormControl component="fieldset">
              <RadioGroup
                aria-label="Hashtage"
                name="hashtagVisibility"
                value={privacy}
                onChange={handleVisibilityChange}
              >
                <FormControlLabel value="public" control={<Radio />} label="Public" />
                <FormControlLabel value="private" control={<Radio />} label="Private" />
              </RadioGroup>
            </FormControl>
          )}
        </Grid>
        <Grid
          container
          direction="row"
          justify="flex-end"
          alignItems="flex-end"
        >
          <Button
            id="submit-button"
            type="submit"
            variant="contained"
            fullWidth
            className={classes.button}
            disabled={loadingGroup || loading}
          >
            POST
            {(loading || loadingGroup) && <CircularProgress size={20} style={{ marginLeft: 5 }} />}

          </Button>
        </Grid>
      </CardBody>
    </form>
  )
}
SubmitPostForm.propTypes = {
  setOpen: PropTypes.func,
  options: PropTypes.array,
  user: PropTypes.object,
}

export default SubmitPostForm
