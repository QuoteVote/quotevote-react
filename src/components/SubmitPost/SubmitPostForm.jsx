import React from 'react'
import TextField from '@material-ui/core/TextField'
import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete'
import { makeStyles } from '@material-ui/core/styles'
import {
  CircularProgress, Divider, Grid, Hidden, IconButton, InputBase, Typography,
} from '@material-ui/core'
import { Controller, useForm } from 'react-hook-form'
import PropTypes from 'prop-types'
import { useMutation } from '@apollo/react-hooks'
import { useDispatch } from 'react-redux'
import { isEmpty } from 'lodash'
import CardBody from '../../mui-pro/Card/CardBody'
import Button from '../../mui-pro/CustomButtons/Button'
import SubmitPostAlert from './SubmitPostAlert'
import { SET_SELECTED_POST } from '../../store/ui'
import { CREATE_GROUP, SUBMIT_POST } from '../../graphql/mutations'

const useStyles = makeStyles({
  root: {
    width: 'auto',
  },
  title: {
    color: '#00CF6E',
    fontSize: 30,
    margin: 'auto',
    paddingLeft: 20,
  },
  exit: {
    color: '#00CF6E',
    fontSize: 30,
    float: 'right',
  },
  text: {
    marginTop: 20,
    fontSize: '20px',
  },
  input: {
    margin: '20px 0px 0px 0px',
  },
  group: {
    margin: '20px 0px',
  },
  groupInput: {
    backgroundColor: 'rgb(160, 243, 204, 0.6)',
    width: 180,
    marginLeft: 20,
  },
  label: {
    color: '#00CF6E',
  },
  button: {
    backgroundColor: '#75E2AF',
    fontSize: 20,
  },
})

const filter = createFilterOptions()

function SubmitPostForm({ options = [], user, setOpen }) {
  const classes = useStyles()
  const dispatch = useDispatch()
  const {
    register, handleSubmit, errors, reset, control,
  } = useForm()
  const [submitPost, { loading }] = useMutation(SUBMIT_POST)
  const [createGroup, { loading: loadingGroup }] = useMutation(CREATE_GROUP)
  const [error, setError] = React.useState(null)

  const onSubmit = async (values) => {
    const { title, text, group } = values
    try {
      let newGroup
      const isNewGroup = group && !('_id' in group)
      if (isNewGroup) {
        newGroup = await createGroup({
          variables: {
            group: {
              creatorId: user._id,
              title: group.title,
              description: `Description for: ${group.title} group`,
              privacy: 'public',
            },
          },
        })
      }
      const postGroupId = isNewGroup ? newGroup.data.createGroup._id : group._id
      const submitResult = await submitPost({
        variables: {
          post: {
            userId: user._id,
            text,
            title,
            groupId: postGroupId,
          },
        },
      })
      const { _id, url } = submitResult.data.addPost
      dispatch(SET_SELECTED_POST(_id))
      setShareableLink(url)
      setShowAlert(true)
    } catch (err) {
      setError(err)
      setShowAlert(true)
    }
  }

  const [shareableLink, setShareableLink] = React.useState('')
  const [showAlert, setShowAlert] = React.useState(false)
  const hideAlert = () => {
    setShowAlert(false)
    setShareableLink('')
    reset()
  }
  const [value, setValue] = React.useState({ title: '', content: '' })
  const [isPasting, setPasting] = React.useState(false)

  const handleTitleChange = (event) => {
    event.persist()
    setValue({ ...value, title: event.target.value })
  }

  const handleContentChange = (event) => {
    event.persist()
    const contentValue = event.target.value
    const validURL = /^(?:http(s)?:\/\/)([\w.-])+(?:[\w.-]+)+([\w\-._~:/?#[\]@!$&'()*+,;=.])+$/
    setValue({ ...value, content: contentValue })
    if (isPasting && contentValue.match(validURL)) {
      parseURLContent(contentValue)
    }
    setPasting(false)
  }

  const parseURLContent = async (url) => {
    try {
      const response = await fetch(`https://cors-anywhere.herokuapp.com/${url}`)
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }
      
      const html = await response.text()
      
      const titleMatch = html.match(/<title[^>]*>([^<]+)<\/title>/i)
      const title = titleMatch ? titleMatch[1].trim() : 'Extracted Content'
      
      const cleanHtml = html
        .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
        .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '')
        .replace(/<[^>]+>/g, ' ')
        .replace(/\s+/g, ' ')
        .trim()
      
      const content = cleanHtml.substring(0, 500) + (cleanHtml.length > 500 ? '...' : '')
      
      if (isEmpty(content)) {
        setError('Could not extract site content.')
        setShowAlert(true)
      } else {
        setValue({ title, content })
      }
    } catch (err) {
      console.error('Error parsing URL:', err)
      setError('Could not extract site content. Please try again.')
      setShowAlert(true)
    }
  }

  const handlePaste = () => {
    setPasting(true)
  }

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {showAlert && (
        <SubmitPostAlert
          hideAlert={hideAlert}
          shareableLink={shareableLink}
          error={error}
          setShowAlert={setShowAlert}
          setOpen={setOpen}
        />
      )}
      <CardBody className={classes.root}>
        <Grid
          container
          direction="row"
        >
          <Typography className={classes.title} variant="body2">Create a Post</Typography>
          <Hidden smDown>
            <IconButton className={classes.exit} onClick={() => setOpen(false)}>X</IconButton>
          </Hidden>
        </Grid>
        <InputBase
          className={classes.input}
          fullWidth
          id="title"
          placeholder="Enter Title"
          value={value.title}
          onChange={(event) => {
            handleTitleChange(event)
          }}
          name="title"
          inputRef={register({
            required: 'Title is required',
          })}
          required
          error={errors.title}
        />
        <Divider />
        <InputBase
          className={classes.input}
          id="text"
          placeholder="Enter text or URL here"
          value={value.content}
          onChange={(event) => {
            handleContentChange(event)
          }}
          onPaste={handlePaste}
          multiline
          fullWidth
          name="text"
          inputRef={register({
            required: 'Post is required',
          })}
          required
          error={errors.text}
        />
        <Divider />
        <Grid
          container
          direction="row"
          justify="flex-start"
          alignItems="center"
          className={classes.group}
        >
          <Typography>
            Who can see your post
          </Typography>

          <Controller
            render={({ onChange, ...props }) => (
              <Autocomplete
                {...props}
                variant="outlined"
                size="small"
                className={classes.groupInput}
                onChange={(event, newValue) => {
                  let data
                  if (typeof newValue === 'string') {
                    data = {
                      title: newValue,
                    }
                  } else if (newValue && newValue.inputValue) {
                    // Create a new value from the user input
                    data = {
                      title: newValue.inputValue,
                    }
                  } else {
                    data = newValue
                  }
                  onChange(data)
                }}
                filterOptions={(groupOptions, params) => {
                  const filtered = filter(groupOptions, params)

                  // Suggest the creation of a new value
                  if (params.inputValue !== '') {
                    filtered.push({
                      inputValue: params.inputValue,
                      title: `Add "${params.inputValue}"`,
                    })
                  }

                  return filtered
                }}
                selectOnFocus
                clearOnBlur
                handleHomeEndKeys
                id="hashtag"
                options={options}
                getOptionLabel={(option) => {
                  // Value selected with enter, right from the input
                  if (typeof option === 'string') {
                    return option
                  }
                  // Add "xxx" option created dynamically
                  if (option.inputValue) {
                    return option.inputValue
                  }
                  // Regular option
                  return option.title
                }}
                renderOption={(option) => option.title}
                renderInput={(params) => (
                  <TextField
                    variant="outlined"
                    className={classes.label}
                    {...params}
                    label="Select a hashtag"
                    name="hashtag"
                    id="hashtag"
                    inputRef={register({
                      required: 'hashtag is required',
                    })}
                  />
                )}
                inputRef={register({
                  required: 'Hashtag is required',
                })}
              />
            )}
            onChange={([, data]) => data}
            name="group"
            control={control}
            defaultValue=""
          />
        </Grid>
        <Grid
          container
          direction="row"
          justify="flex-end"
          alignItems="flex-end"
        >
          <Button
            id="submit-button"
            type="submit"
            variant="contained"
            fullWidth
            className={classes.button}
            disabled={loadingGroup || loading}
          >
            POST
            {(loading || loadingGroup) && <CircularProgress size={20} style={{ marginLeft: 5 }} />}

          </Button>
        </Grid>
      </CardBody>
    </form>
  )
}

SubmitPostForm.propTypes = {
  setOpen: PropTypes.func,
  options: PropTypes.array,
  user: PropTypes.object,
}

export default SubmitPostForm

