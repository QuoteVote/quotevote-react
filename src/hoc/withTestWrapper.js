import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <Router history={hist}>
            <Component {...props} />
          </Router>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
