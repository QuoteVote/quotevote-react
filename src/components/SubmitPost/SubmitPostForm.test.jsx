import React from 'react'
import { render } from '@testing-library/react'

// Component being tested
import SubmitPostForm from './SubmitPostForm'
import withTestWrapper from '../../hoc/withTestWrapper'

const SubmitPostFormWrapper = withTestWrapper(SubmitPostForm)

describe('SubmitPostForm test -', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SubmitPostFormWrapper />,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
