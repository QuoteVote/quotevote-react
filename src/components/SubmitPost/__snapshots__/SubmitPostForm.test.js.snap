// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitPostForm test - renders correctly 1`] = `
<form>
  <div
    class="makeStyles-cardBody-10 makeStyles-root-1"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <p
        class="MuiTypography-root makeStyles-title-2 MuiTypography-body2"
      >
        Create a Post
      </p>
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-exit-3"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          X
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiInputBase-root makeStyles-input-5 MuiInputBase-fullWidth"
      required=""
    >
      <input
        class="MuiInputBase-input"
        id="title"
        name="title"
        placeholder="Enter Title"
        required=""
        type="text"
        value=""
      />
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiInputBase-root makeStyles-input-5 MuiInputBase-fullWidth MuiInputBase-multiline"
      required=""
    >
      <textarea
        class="MuiInputBase-input MuiInputBase-inputMultiline"
        id="text"
        name="text"
        placeholder="Enter text or URL here"
        required=""
        rows="1"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInputBase-inputMultiline"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
        tabindex="-1"
      />
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiGrid-root makeStyles-group-6 MuiGrid-container MuiGrid-align-items-xs-center"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Who can see your post
      </p>
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root makeStyles-groupInput-7 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        role="combobox"
        variant="outlined"
      >
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-label-8 MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="hashtag"
            id="hashtag-label"
          >
            Select a hashtag
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="hashtag"
              name="hashtag"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-19 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-21"
              >
                <span>
                  Select a hashtag
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-end MuiGrid-justify-xs-flex-end"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-23 makeStyles-fullWidth-24 makeStyles-button-9"
        id="submit-button"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          POST
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</form>
`;
